src/meta/processors/job/BalanceJobExecutor.cpp

```diff
@@ -665,174 +439,327 @@ void BalanceJobExecutor::calDiff(const HostParts& hostParts,
-  if (!nebula::ok(activeHostsRet)) {
-    auto retCode = nebula::error(activeHostsRet);
-    LOG(ERROR) << "Get active hosts failed, error: " << apache::thrift::util::enumNameSafe(retCode);
-    return retCode;
+  for (auto& zoneMapEntry : lostZones) {
+    Zone* zone = zoneMapEntry.second;
+    for (auto& hostMapEntry : zone->hosts_) {
+      for (PartitionID partId : hostMapEntry.second.parts_) {
+        HostAddr dst = chooseZoneToInsert(partId);
+        tasks.emplace_back(
+            jobId_, spaceInfo_.spaceId_, partId, hostMapEntry.first, dst, kvstore_, adminClient_);
+      }
+      hostMapEntry.second.parts_.clear();
+    }
+    zone->calPartNum();
   }
-  auto hosts = std::move(nebula::value(activeHostsRet));
 
-  if (hosts.empty()) {
-    LOG(ERROR) << "There is no active hosts";
-    return nebula::cpp2::ErrorCode::E_NO_HOSTS;
+  int32_t totalPartNum = 0;
+  int32_t avgPartNum = 0;
+  for (auto& z : sortedActiveZones) {
+    totalPartNum += z->partNum_;
   }
-  lostHosts_.reserve(paras_.size() - 1);
-  for (size_t i = 0; i < paras_.size() - 1; i++) {
-    lostHosts_.emplace_back(HostAddr::fromString(paras_[i]));
+  avgPartNum = totalPartNum / sortedActiveZones.size();
+  int32_t remainder = totalPartNum - avgPartNum * sortedActiveZones.size();
+  int32_t leftBegin = 0;
+  int32_t leftEnd = 0;
+  int32_t rightBegin = 0;
+  int32_t rightEnd = sortedActiveZones.size();
+  for (size_t i = 0; i < sortedActiveZones.size(); i++) {
+    if (avgPartNum <= sortedActiveZones[i]->partNum_) {
+      leftEnd = i;
+      break;
+    }
   }
-  return nebula::cpp2::ErrorCode::SUCCEEDED;
-}
-
-nebula::cpp2::ErrorCode DataBalanceJobExecutor::finish(bool ret) {
-  std::lock_guard<std::mutex> lg(lock_);
-  return finishInternal(ret);
-}
-
-nebula::cpp2::ErrorCode DataBalanceJobExecutor::finishInternal(bool ret) {
-  CHECK(!lock_.try_lock());
-  plan_.reset(nullptr);
-  running_ = false;
-  auto rc = onFinished_(ret);
-  if (rc != nebula::cpp2::ErrorCode::SUCCEEDED) {
-    return rc;
+  for (size_t i = leftEnd; i < sortedActiveZones.size(); i++) {
+    if (avgPartNum < sortedActiveZones[i]->partNum_) {
+      rightBegin = i;
+      break;
+    }
   }
-  return ret ? nebula::cpp2::ErrorCode::SUCCEEDED : nebula::cpp2::ErrorCode::E_BALANCER_FAILURE;
-}
-
-folly::Future<Status> DataBalanceJobExecutor::executeInternal(HostAddr&& address,
-                                                              std::vector<PartitionID>&& parts) {
-  UNUSED(address);
-  UNUSED(parts);
-  std::unique_lock<std::mutex> lg(lock_);
-  if (!running_) {
-    if (plan_ == nullptr) {
-      auto retCode = buildBalancePlan();
-      if (retCode != nebula::cpp2::ErrorCode::SUCCEEDED) {
-        if (retCode == nebula::cpp2::ErrorCode::E_BALANCED) {
-          finishInternal(true);
-          return Status::OK();
-        } else {
-          return Status::Error(apache::thrift::util::enumNameSafe(retCode));
+  for (int32_t right = rightBegin; right < rightEnd;) {
```

Comment:
We need to check if the plan is valid and if it breaks any invariant:
1. All parts in removed zones are gone
2. A part can't have more than 1 in the same zone.